workspaces/widgets.py

```diff
@@ -60,3 +80,47 @@ def get_current_workspace(user: AppUser, session: dict) -> "Workspace":
 
 def set_current_workspace(session: dict, workspace_id: int):
     session[SESSION_SELECTED_WORKSPACE] = workspace_id
+
+
+def get_next_route_fn(
+    route_fn: typing.Callable, workspace: Workspace
+) -> typing.Callable:
+    """
+    When we need to redirect after user changes the workspace.
+    """
+    from routers.account import (
+        account_route,
+        profile_route,
+        workspaces_route,
+        workspaces_members_route,
+    )
+
+    if workspace.is_personal and route_fn in (
+        workspaces_members_route,
+        workspaces_route,
+    ):
+        # personal workspaces don't have a members page
```

Comment:
This redirect needs to be inside these routes themselves, not here

Reply:
The problem is that we end up with two sources of truth - the `session_state` and the URL path params.

e.g. you're on `/workspaces/devs-team-workspace-0A/saved/` and change the workspace from this selector to `Personal Account`.

When rendering workspace `saved` tab, we will see that the URL belongs to the workspace. We also end up at this same state when I share a `/workspaces/devs-team-workspace-0A/saved/` link with you were on `Personal Account` according to `gui.session_state`.

Reply:
The intended behaviour is first case is that your `Personal Account` is selected and redirected to `/account/saved`, and in the second is that the URL overrides `gui.session_state`.

I created this other redirect path so that we can deal with the first case differently.

Reply:
So the URL path params will always override `gui.session_state` if they have a conflict.

The selector widget can redirect / link to the other workspace's URL if a change needs to be made.